### PR TITLE
Restructure sidebar markup for consistency and control over color

### LINKED ISSUE #531: Fix sidebar collapse toggle z-index
**tl;dr**: Change markup so toggle is stacked on top of main; or give the toggle a `z-index` to pull it above the main content area layers.

Unfortunately, I think we do need `z-index` or another solution in some capacity. In particular, we need the collapse toggle to have a higher `z-index` than main so that nested sidebar layouts work. Without `z-index` ordering, the inner sidebar layout covers the collapse toggle of the outer sidebar.

You can see the problem in our docs now in the [Nested Sidebars](https://rstudio.github.io/bslib/articles/sidebars.html#nested-sidebars) section.

_Originally posted by @gadenbuie in https://github.com/rstudio/bslib/issues/509#issuecomment-1482907724_

Another place where I've noticed this is with leaflet maps in the main sidebar area. (You can currently see this in [the final sidebar example](https://rstudio.github.io/bslib/articles/sidebars.html#styling-and-customization) in our docs.) The collapse toggle is hidden behind the leaflet map, but giving the toggle z-index of even 1 raises the toggle above the widget.

| No z-index | `z-index: 1` |
|:---:|:--:|
| image | image |

Now that I've written it out, I think the fix might actually be quite simple: to move the toggle in the markup so that it's added _after_ the main content area. In the example above, it's not leaflet in particular; any filling item in the sidebar area would cover the toggle.

### PR DESCRIPTION
**Pull request by gadenbuie**
For #500
From #516 (partially)
Fixes #531

## PR Updates

The main goal of this PR was to adjust the collapse toggle so that its colors match the colors of the sidebar. This uncovered a number of smaller issues that resulted in rather large changes under-the-hood.

1. Both `layout_sidebar()` and `sidebar()` gained `fg` and `bg` arguments. In both cases, if only one is provided we choose whichever of `$black` or `$white` has higher contrast for the other color.

   image

3. If `sidebar()` has a background color, we also set the sidebar layout container's background color. This makes the collapsing animation seamless (otherwise the sidebar disappears abruptly and we animate blank space).

    https://user-images.githubusercontent.com/5420529/229906145-06d0d001-4cbd-4d36-a2e0-5cb526ae83ed.mp4

5. To accommodate the last point,`sidebar()` now includes its colors in the data structure so that `layout_sidebar()` can take them into account. If they are set, the `layout_sidebar()` sets the main content area colors to match the body colors.

7. While working on the collapse toggle, I tackled #531. To ensure consistent stacking order, we now always include the sidebar elements in the following order: main, sidebar, collapse toggle. Rather than moving them in the markup, we adjust their position in the CSS.

   https://user-images.githubusercontent.com/5420529/229906671-894c4999-d8bd-4d06-8071-10ae0ca0c933.mp4

8. The border between main/sidebar areas is now applied to the sidebar. **Update:** with the new sidebar markup, the outer border of the layout won't receive the sidebar color.

10. While working on the border color, I also moved around our border-radius rules. The final effect is the same.

11. Lastly, after adjusting the markup and CSS, I realized that we can also move the grid column logic into the CSS. In doing this, I removed `--bslib-sidebar-columns` and `--bslib-sidebar-columns-collapsed` and replaced them with a `--bslib-sidebar-width` CSS variable. I also ensured that the sidebar is never wider than the parent container. The end result is that everything we set in the attributes or CSS of the sidebar layout relate to state or user preference.

12. The collapse toggle buttons now maintain their position through all states on mobile. This is consistent with sidebar toggles on desktop. In other words, the second-level sidebar is always second from the right on mobile.
     

14. The sidebar now defaults to `$gray-100`. This default is set via the `$bslib-sidebar-bg` SASS variable, which could be set at the theme level, in addition to `$bslib-sidebar-fg`. The sidebar background and foreground are still locally configurable via the `bg` and `fg` arguments described at the top of this PR.
   image

## Examples

Among other examples, I've been using the code below to generate many card variations for testing and review.

<details>
<summary>Example</summary>

```r
library(bslib)

colors <- list(
  list(color = "#FFFAFA", name = "Snow"),
  list(color = "#B22222", name = "Fire Brick"),
  list(color = "#2F4F4F", name = "Dark Slate Gray"),
  list(color = "#556B2F", name = "Dark Olive Green"),
  list(color = "#483D8B", name = "Dark Slate Blue"),
  list(color = "#00FA9A", name = "Medium Spring Green"),
  list(color = "#8B4513", name = "Saddle Brown"),
  list(color = "#FFDAB9", name = "Peach Puff"),
  list(color = "#228B22", name = "Forest Green"),
  list(color = "#FFD700", name = "Gold"),
  list(color = "#8B008B", name = "Dark Magenta"),
  list(color = "#FF8C00", name = "Dark Orange"),
  list(color = "#FF69B4", name = "Hot Pink"),
  list(color = "#4169E1", name = "Royal Blue"),
  list(color = "#00CED1", name = "Dark Turquoise"),
  list(color = "#F0E68C", name = "Khaki")
)

sidebars <- lapply(colors, function(x) {
  do_main_bg <- runif(1) > 0.5
  do_border <- runif(1) > 0.5
  do_border_radius <- runif(1) > 0.5
  do_border_color <- FALSE # runif(1) > 0.75
  sb_width <- paste0(floor(runif(1, 150, 500)), "px")
  sb_pos <- sample(c("left", "right"), 1)
  as_card <- runif(1) > 0.5
  state <- sample(c(TRUE, FALSE, NA), 1)

  ret <- layout_sidebar(
    sidebar = sidebar(
      position = sb_pos,
      width = sb_width,
      open = state,
      bg = if (prismatic::clr_extract_lightness(x$color) > 50) {
        as.character(prismatic::clr_darken(x$color, 0.1))
      } else {
        as.character(prismatic::clr_lighten(x$color, 0.1))
      },
      HTML(sprintf(
        "Sidebar: %s<br>position: %s<br>border: %s<br>radius: %s<br>width: %s<br>open: %s",
        x$name, sb_pos, do_border, do_border_radius, sb_width,
        if (is.na(state)) "always" else if (state) "open" else "closed"
      ))
    ),
    height = "350px",
    bg = if (do_main_bg) x$color,
    border = do_border,
    border_radius = do_border_radius,
    border_color = if (do_border_color) "#00CED1",
    lorem::ipsum(1, 2)
  )

  if (!as_card) htmltools::tagAppendAttributes(ret, class = "my-2") else {
    card(
      class = "my-2",
      if (runif(1) > 0.25) card_header("A Card"),
      ret
    )
  }
})

extra_sidebars <- list(
  layout_sidebar(
    height = "300px",
    sidebar = sidebar("Default sidebar"),
    p("Default content"),
    lorem::ipsum(1, 2)
  ),
  layout_sidebar(
    sidebar(
      width = "33%",
      fg = "#9B59B6",
      p("width: 33%")
    ),
    lorem::ipsum(1, 2)
  ),
  layout_sidebar(
    fillable = TRUE,
    class = "p-0",
    sidebar(
      width = "20%",
      bg = "#FF7847",
      fg = "",
      "Left Sidebar 1"
    ),
    layout_sidebar(
      fillable = TRUE,
      class = "p-0",
      border = FALSE,
      border_radius = FALSE,
      sidebar(
        width = "30%",
        bg = "#FF70A6",
        position = "right",
        "Right Sidebar"
      ),
      layout_sidebar(
        border = FALSE,
        border_radius = FALSE,
        sidebar(
          width = "40%",
          bg = "#FFB499",
          "Left Sidebar 2",
        ),
        lorem::ipsum(3, 1)
      )
    )
  )
)

page_fluid(
  layout_column_wrap(
    width = "400px",
    fill = TRUE,
    !!!extra_sidebars,
    !!!sidebars
  )
)

page_fixed(
  div(
    class = "my-5",
    !!!extra_sidebars,
    !!!sidebars
  )
)
```

</details>

### CODE REVIEW

File: R/sidebar.R [hunk 1]
```diff
@@ -199,6 +220,10 @@ sidebar_close <- function(id, session = get_current_session()) {
   session$onFlush(callback, once = TRUE)
 }
 
+collapse_icon <- function() {
+  HTML("<svg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 16 16' class='collapse-icon' fill='currentColor'><path fill-rule='evenodd' d='M1.646 4.646a.5.5 0 0 1 .708 0L8 10.293l5.646-5.647a.5.5 0 0 1 .708.708l-6 6a.5.5 0 0 1-.708 0l-6-6a.5.5 0 0 1 0-.708z'/></svg>")
+}
```
Comment (cpsievert) [inline R/sidebar.R:235]: Maybe worth using `bsicons` if available?

Comment (cpsievert) [reply R/sidebar.R:235]: If we do, maybe we should also be doing that for `card(full_screen = TRUE)`

File: R/sidebar.R [hunk 2]
```diff
@@ -47,15 +48,27 @@ sidebar <- function(
...
 
   collapse_tag <- tags$button(
     class = "collapse-toggle",
     type = "button",
     title = "Toggle sidebar",
-    style = css(display = if (hide_collapse) "none"),
+    style = css(
+      display = if (hide_collapse) "none",
+      background_color = bg,
+      color = fg
```
Comment (cpsievert) [inline R/sidebar.R:67]: Could this just inherit from the parent?

File: R/sidebar.R [hunk 3]
```diff
@@ -65,14 +78,16 @@ sidebar <- function(
       class = c("sidebar", class),
       style = css(
         background_color = bg,
-        color = if (!is.null(bg)) get_color_contrast(bg)
+        color = fg
```
Comment (cpsievert) [inline R/sidebar.R:81]: Could this just inherit from the parent?

File: inst/components/sidebar.scss [hunk 4]
```diff
@@ -60,12 +60,11 @@ $bslib-sidebar-border: var(--bs-card-border-width, #{$card-border-width}) solid
   }
 
   > .collapse-toggle {
-    grid-row: 1 / 2;
     grid-column: 1 / 2;
```
Comment (cpsievert) [inline inst/components/sidebar.scss:63]: Is this change necessary?

File: inst/components/sidebar.scss [hunk 5]
```diff
@@ -13,8 +13,9 @@ $bslib-sidebar-border: var(--bs-card-border-width, #{$card-border-width}) solid
   --bslib-collapse-toggle-right-transform: -90deg;
 
   display: grid !important;
-  // Should always be defined as an inline style on the layout_sidebar() container
-  grid-template-columns: var(--bslib-sidebar-columns);
+  grid-template-columns:
+    Min(calc(100% - #{$bslib-sidebar-icon-size}), var(--bslib-sidebar-width, 250px))
+    minmax(0, 1fr);
```
Comment (cpsievert) [inline inst/components/sidebar.scss:18]: Let's consider putting each column definition into a CSS variable (defined above), for example:


```css
---bslib-sidebar-width-actual: Min(calc(100% - #{$bslib-sidebar-icon-size}), var(--bslib-sidebar-width, 250px))
```

then when we need it:

```css
grid-template-columns: var(---bslib-sidebar-width-actual) var(---bslib-main-column)
```

Comment (gadenbuie) [reply inst/components/sidebar.scss]: Fixed in 86afeec9710ea27459c6129f2fdef15188d5ef11

I went with two sass variables, `$bslib-sidebar-column-sidebar` and `$bslib-sidebar-column-main`. Overall a definite improvement! I stuck with SASS variables because we don't need to expose these client-side. The sidebar column width carries the `--bslib-sidebar-width` variable, so that can be used client side to adjust sidebar width (for the adventurous slash if needed).

File: inst/components/sidebar.scss (hunk 4 above)
Comment (gadenbuie) [reply inst/components/sidebar.scss:64]: Done in b9ecbe649ac966716e530cd7cb67147a360bf3d7

File: R/sidebar.R (hunk 2 above)
Comment (gadenbuie) [reply R/sidebar.R]: I tackled this in e8c991f1ae673e2fff681ddd6d6d3027059275cc. I ended up wiring up a new SASS variable to help theme the default sidebar background color. (See point 9 in the main PR description.)

File: R/sidebar.R (hunk 1 above)
Comment (gadenbuie) [reply R/sidebar.R]: Done in d8e5b0577f1b4f90b425c8b237fd950171168310. For now, I only used bsicons for the collapse icon. I actually like this approach a lot and if we decided to make the collapse and fullscreen icons (more easily) customizable, bsicons would be a nice way to achieve this.

File: inst/components/sidebar.scss [hunk 6]
```diff
@@ -1,8 +1,11 @@
 $bslib-sidebar-padding: $spacer * 1.5 !default;
 $bslib-sidebar-icon-size: $spacer !default;
-
+$bslib-sidebar-bg: var(--bslib-sidebar-bg, #{$gray-200}) !default;
```
Comment (cpsievert) [inline inst/components/sidebar.scss:3]:
```suggestion
$bslib-sidebar-bg: $gray-200 !default;
$bslib-sidebar-fg: color-contrast($bslib-sidebar-bg) !default;
```

Comment (gadenbuie) [reply inst/components/sidebar.scss:3]: I'm going to have to come back to this next week. If we set the background color of the sidebar, then we need to set the parent layout container colors to get the right animation. And then we need to set the main content area colors by default too. In which case, our problem would then shift to the R side where we'd need a way to unset everything such that the whole sidebar layout could inherit from its parent.

In other words, I need to think about this some more.

File: inst/components/sidebar.scss [hunk 7]
```diff
@@ -1,8 +1,11 @@
...
 $bslib-sidebar-border: var(--bs-card-border-width, #{$card-border-width}) solid var(--bs-card-border-color, #{$card-border-color}) !default;
 
+$bslib-sidebar-column-sidebar: Min(calc(100% - #{$bslib-sidebar-icon-size}), var(--bslib-sidebar-width, 250px));
+$bslib-sidebar-column-main: minmax(0, 1fr);
+
 .bslib-sidebar-layout {
```
Comment (cpsievert) [inline inst/components/sidebar.scss:9]: If we do move forward with the gray-200 (or whatever), let's make sure it looks good with `actionButton()` (and other shiny inputs). Maybe we can set a CSS variables to tweak some styles if need be

File: inst/components/sidebar.scss (hunk 6 above)
Comment (gadenbuie) [reply inst/components/sidebar.scss]: I applied the above change, which now works with the updated sidebar markup

File: R/sidebar.R [hunk 8]
```diff
@@ -96,47 +110,60 @@ layout_sidebar <- function(
...
+  if (is.null(bg) && !is.null(fg)) {
+    bg <- get_color_contrast(fg)
+  }
+  # 2. If the sidebar has a color, reset main to body fg/bg color
+  if (is.null(fg) && !is.null(sidebar$fg)) {
+    fg <- "var(--bs-body-color)"
+  }
+  if (is.null(bg) && !is.null(sidebar$bg)) {
+    bg <- "var(--bs-body-bg)"
+  }
```
[the remark this answers is not in the document]
Comment (gadenbuie) [reply R/sidebar.R]: This wasn't necessary anymore after the sidebar markup changes

File: inst/components/sidebar.scss (hunk 7 above)
Comment (gadenbuie) [reply inst/components/sidebar.scss:10]: I went with `$gray-100` for the default, which works nicely out of the box for action buttons and other elements

**Comment by gadenbuie** [comment]
**Update:** I simplified a lot of the color logic by updating the sidebar markup. We now use `.sidebar` as a wrapper container and place all of the sidebar contents inside a child `.sidebar-wrapper` div. This lets us take hide the sidebar content but leave the wrapper element as we transition the grid column size to collapse or expand the sidebar.

This gives us decent sidebar animations, even with the new default background color.

https://user-images.githubusercontent.com/5420529/231277387-fefd1397-9ee8-4cba-85c2-4db572f778a3.mp4

With this PR the layout markup is now

```
div.bslib-sidebar-layout
  div.main
  div.sidebar
    div.sidebar-content
  button.collapse-toggle
```

For consistency with the previous markup we could use `form.sidebar-content`, but after some research I don't think we need that element to be a `<form>` element. We'd only need the `<form>` if the input values were going to be submitted via the browser; in the majority of cases inputs in sidebars will be managed via JavaScript.

File: R/sidebar.R [hunk 9]
```diff
@@ -199,6 +218,14 @@ sidebar_close <- function(id, session = get_current_session()) {
   session$onFlush(callback, once = TRUE)
 }
 
+collapse_icon <- function() {
+  if (!rlang::is_installed("bsicons")) {
```
Comment (gadenbuie) [inline R/sidebar.R:222]: Use internal `is_available()`

```suggestion
  if (!is_available("bsicons")) {
```
